core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -650,17 +651,55 @@ private void applyInitializingShard(final ClusterState state, final IndexMetaDat
             final RecoveryState recoveryState = new RecoveryState(indexShard.shardId(), shardRouting.primary(),
                 RecoveryState.Type.STORE,
                 nodes.getLocalNode(), nodes.getLocalNode());
-            indexShard.markAsRecovering("from store", recoveryState); // mark the shard as recovering on the cluster state thread
-            threadPool.generic().execute(() -> {
-                try {
-                    if (indexShard.recoverFromStore(nodes.getLocalNode())) {
-                        shardStateAction.shardStarted(shardRouting, "after recovery from store", SHARD_STATE_ACTION_LISTENER);
+            Index mergeSourceIndex = indexShard.indexSettings().getIndexMetaData().getMergeSourceIndex();
+            if (mergeSourceIndex != null) {
+                IndexService sourceIndexService = indicesService.indexService(mergeSourceIndex);
```

Comment:
do we need to protect against the index not being there? right now it will blow up on the cluster state update thread with an NPE? we assume this primitive will be called when the source shard is assigned to the same node, but we either need to enforce somehow or protect properly against this not being the case.